### PR TITLE
Refactor implementation

### PR DESCRIPTION
**Pull request by ekohl**
This is a work in progress refactor of the code base. The goal is to make the code suitable for multiple providers by pushing all PHP IPAM handling into that provider. It's entirely untested and not well sorted (as can be seen by various fixup commits). Sadly the tests that exist all fail in master already so there's not a lot to go on.

I'm pushing this now as a first draft and would like some feedback on the general idea. The goal is to move to a well defined API with a clean separation from the actual provider.

### CODE REVIEW

**Comment by grizzthedj** [comment]
There is no question that this code needs a serious refactoring. I appreciate your efforts here. I have always envisioned this plugin working with multiple providers and have been wanting to introduce the provider design for a while now(you can see in the README). Between the additional scope being added(i.e. IPv6, dual stack, IP cache etc.) and battling with other projects/fires it has been difficult to find the time. 

My focus has been to get the phpIPAM integration working first(with Forman patch merged), then refactor in the provider design once I started work on the next provider. That was going to be Netbox in Q3 this year, which I still need to do. 

I would be happy to assist with this and testing, once ready.

File: lib/smart_proxy_ipam/phpipam/phpipam_helper.rb
Comment (m4c3) [inline lib/smart_proxy_ipam/phpipam/phpipam_helper.rb]: typo: emtpy? -> empty?

**Review by m4c3 — CHANGES_REQUESTED** [review]

File: lib/smart_proxy_ipam/phpipam/phpipam_helper.rb
Comment (m4c3) [inline lib/smart_proxy_ipam/phpipam/phpipam_helper.rb]: array.filter is only available since ruby 2.6
array.filter is an alias for array.select

**Comment by ekohl** [comment]
I added those fixes, but this needs a lot of work to finish and I don't know if I can commit to that. I'd be happy if someone else took over, whether that's with my commits or a fresh start.

The goal of this PR was to separate the API implementation from the actual provider. It hasn't done that yet, but the next step would have been to implement the dependency injection part which allows swapping the provider via config files.

**Comment by grizzthedj** [comment]
@ekohl - It has been my intention to do this from the start, so would be happy to take over where you left off in your branch. I meant to get to this, however the workload has nearly doubled since covid, so it has been difficult. I will try to get to it next week as the interest seems to be growing. 

Question: How will Foreman tell the Plugin which provider to pull an IP from? If someone uses both Netbox and phpIPAM, for example, there should be a way to associate a Foreman subnet with an IPAM provider.

**Comment by m4c3** [comment]
to be consistend with dns and dhcp proxys, the user has to setup multiple foreman proxys to use different ipams

**Comment by grizzthedj** [comment]
@ekohl @lzap - are there any concerns with supporting multiple IPAM providers in this plugin? In our use case, we need a single proxy to be able to talk to more than one IPAM provider(I.e. phpIPAM and Netbox to be specific). Swapping configs to change provider does not really allow us to do this. I think there is some value(for everyone) in the proxy(plugin) having the flexibility to be able to pull ip’s from multiple providers, depending on the selected provider in foreman. 

We can add an ExternalIpamProvider model in Foreman, which can be selected, when creating External IPAM subnets in Foreman. This ExternalIpamProvider drop down would be shown for External IPAM subnets only. 

I realize that the code base in its current state would make this difficult, but I plan on fixing this ASAP so that providers can added easily. Hopefully will be done this end of next week. Please let me know your thoughts on this approach.

**Comment by lzap** [comment]
> are there any concerns with supporting multiple IPAM providers in this plugin

Not at all, I would actually like this repo to be merged into smart proxy core once it stabilizes. The only reason why I would like to see it incubating outside is supportability - once we merge it we need to start supporting it, that includes our products Red Hat Satellite and Orachino. So I would like to be sure the code is mature enough.

I haven't read the patch yet, however before I do let me say that you can borrow most of the code from smart proxy core, any module that has providers would do it. Good example would be DHCP and its three implementations

> having the flexibility to be able to pull ip’s from multiple providers, depending on the selected provider in foreman.

Does the current design block some workflows? You need to deploy multiple smart-proxies per provider in this model, but that's what we've been encouraging users to do for any kind of smart proxy service. The proxy itself is a tiny process.

**Comment by grizzthedj** [comment]
@lzap @ekohl I have refactored the smart_proxy_ipam plugin implementation with the following changes. I'm not completely finished yet, but most of the core implementation is done, and would like to get some feedback on it. I don't think this new implementation blocks any workflows.

  1. Separated the API from provider implementation. This greatly simplifies the adding of new providers
  2. Consistent use of error codes and exception handling and JSON parsing(this was inconsistent before)
  3. Extracted the temporary IP cache out to a separate class, which can be used by each provider
  4. Moved base HTTP functions to separate class(still need to work the auth functions more)
  5. More robust validation
  6. Added better tests
  7. Added rubocop

Basically, this new implementation allows a single Smart Proxy to support multiple IPAM providers. For example, to get the next available IP for phpIPAM, Foreman would make this API call: 

  `/phpipam/subnet/:address/:prefix/next_ip`

For Netbox it would be: 

  `/netbox/subnet/:address/:prefix/next_ip`

On the Foreman side, I propose adding an `ExternalIpamProvider` model(a single "name" column), which can be selected via dropdown when creating External IPAM subnets in Foreman. This drop down would be shown for External IPAM subnets only. When foreman makes requests to the API, it will use the subnet's IPAM provider string as the first REST parameter in the path(as per the examples above). When support for new providers are added, on the foreman side, it should be as simple as adding another record to the `ExternalIpamProvider` table.

Also, there is a PR from @m4c3 for adding Netbox support which I have started testing. So far so good there, just need to add the group(aka Tenant in Netbox) functions and should be good to go. I will need to create a patch for foreman to update API documentation, and make use of response codes etc. I will be starting to test this new implementation with Foreman likely next week. Once everything is good there, I will merge the Netbox PR and the new implementation. 

The new implementation is in the [refactor/providers](https://github.com/grizzthedj/smart_proxy_ipam/tree/refactor/providers) branch. Let me know your thoughts on the new implementation, and the proposed approach in general.

**Comment by lzap** [comment]
Hey,

It's great to see this moving. 

> /phpipam/subnet/:address/:prefix/next_ip

Why do you have implementation name in the URL? This should be completely transparent to API clients (e.g. Foreman). You set up one of the implementations in the proxy configuration and the API remains common across all of them. I know it can be challenging to find a common data model but this really pays off - smart proxy API has been stable for more than 10 years for sevices like DHCP or DNS. My advice is: keep it simple. If there are things that only some providers supports, then I suggest to have it as an optional data field.

This is all my fault, I should have explained you how things are implemented in the DHCP module. In general, the API is generic:

https://github.com/theforeman/smart-proxy/blob/develop/modules/dhcp/dhcp_api.rb

As you can see, it hands requests over to "server" instance which is the implementation instance. It have some shared code that each implementation use (essentially it holds records in memory):

https://github.com/theforeman/smart-proxy/blob/develop/modules/dhcp_common/server.rb

And then there is implementation of the server which overrides some methods:

https://github.com/theforeman/smart-proxy/blob/develop/modules/dhcp_common/isc/omapi_provider.rb

Everything is tied together via configuration, in the main configuration file you can select which provider you want to use:

https://github.com/theforeman/smart-proxy/blob/develop/config/settings.d/dhcp.yml.example

And then each provider has its own configuration file with more details (e.g. URL to the service, username, password):

https://github.com/theforeman/smart-proxy/blob/develop/config/settings.d/dhcp_isc.yml.example

Since we would like to merge this into smart-proxy core in the long-term it would be great to follow the same design.

**Comment by ekohl** [comment]
Like what @lzap said: the idea is that the Smart Proxy provides a single API so Foreman doesn't have to care about the exact provider. This encapsulation keeps the Foreman code base more straight forward.

This PR was working towards isolating the implementations. The next step would be to implement actual providers to the API.

Edit: submitted too soon. In particular, ad2b8672e706208e420cf219f30192c937e081b1 makes the first big step towards this. It introduces a provider class. The next step would be to implement the dependency injection that Smart Proxy provides.

**Comment by ekohl** [comment]
I've rebased this and dealt with most fixup commits. The whole point is that `lib/smart_proxy_ipam/phpipam/phpipam_api.rb` should end up as `lib/smart_proxy_ipam/api.rb` and knows **nothing** about phpipam. Then there should be dependency injection to inject *a* provider. Removing all phpipam specifics was, as you can see, a huge job. Even then in the actual REST API a lot of details leak into it.

I've seriously considered starting from scratch. The problem is that the code is already merged in Foreman core (which I was strongly against) so now there's a painful migration ahead.

**Comment by grizzthedj** [comment]
@ekohl @lzap 

As I mentioned, this branch is still a WIP, so I'm not finished yet.  

I will be looking to complete this refactoring and add the dependency injection for the provider, in the coming weeks.

**Comment by ekohl** [comment]
I took a bit of time to work on this. The current version has dependency injection and can load a different provider. Since the current tests in master all fail I wrote a new integration test that at least loads everything. The README is updated with the new config files and how to use it. As a bonus, I added input validation on the phpipam connection parameters. The provider now refuses to load if the settings are invalid.

There is still a bit of leakage in the layers. In particular `get_subnet` should return nil or raise an exception. Then `check_subnet_exists!` can be removed.

I also think that `provider` that calls `authenticated?` should be investigated. Perhaps this could also live in `load_dependency_injection_wirings` where it instantiates the class.

Another possible improvement is using https://github.com/theforeman/smart-proxy/blob/develop/lib/proxy/validations.rb for validations. It can use `validate_mac` from that. `validate_ip` should not be used since it only supports IPv4.

However, all in all I think this branch is getting pretty close. What should now be done is see if you can implement some `externalipam_netbox` provider.

It should also be verified whether the Foreman side can deal with the API modifications I made. I chose to return HTTP 201 when a resource was created etc.

**Comment by lzap** [comment]
Nice work, it's great to see this closing the gap. Foreman desperately needs a good IPAM integration.

**Comment by grizzthedj** [comment]
@ekohl @lzap I'm having some trouble with the DI wirings. Getting the following error, when trying to hit any endpoint: 

```
{"error": "Dependency 'externalipam_client' is undefined"}
```

Does any code need to be added to the proxy core to make this work? I'm assuming not, but just thought I would check. 

The integration test itself is green, however the dependency is undefined at runtime. Any ideas?

**Comment by ekohl** [comment]
@grizzthedj the plugin / provider (depending on what it exactly is) needs to be loaded. That's what `load_classes` is for. However, in the test suite often various files are already loaded from a previous test. I think the line has a bug:

```ruby
      require 'smart_proxy_ipam/phpipam/phpipam_client.rb'
```

I probably should not include the `.rb` extension there. Pushing an update.

**Comment by grizzthedj** [comment]
@ekohl That may have been an issue, but still seeing the same error. It looks like the `load_classes` and `load_dependency_injection_wirings` methods are not even being called from anywhere, which is why the PhpipamClient class is not getting loaded. This is in both `ipam.rb` and `phpipam_plugin.rb`. Where are these methods called from? Proxy core?

**Comment by ekohl** [comment]
I think I was a bit too quick with my patches and I think this has a better chance of working.

**Comment by grizzthedj** [comment]
@ekohl The DI is working now. Thanks for updating!

**Comment by grizzthedj** [comment]
@ekohl @lzap I have incorporated most of these changes, from this PR, into develop. I'm not going to merge this PR itself, since I had already started refactoring on my own branch, and it was easiest for me to just cherry pick the changes from your branch into mine. Hope you are OK with this @ekohl. 

I have also integrated @m4c3's work on the Netbox provider(thanks again @m4c3!). I changed the Netbox group implementation from "Tenant" to "VRF". The Tenant component does not allow us to properly namespace ip's and prefixes. There are more specific details in this GitHub issue I posted a little over a month ago: https://github.com/netbox-community/netbox/issues/5018

I have also created a simple rake task to create the necessary files(with method stubs) required for implementing a new IPAM provider. The `external_<PROVIDER>.yml` needs to be created manually in the proxy core settings.d directory though. This was probably overkill, but just wanted to make it as easy as possible to add new providers.

```
bundle exec rake add_provider["netbox"]
```

There is still more work to be done, which I will be continuing in the coming weeks, however I believe the code is in a good state now with separation of the API from the different provider implementations. The dependency injection is also working.

Next steps:

1. Regression Test all External IPAM actions using Foreman, for both providers(**This is currently in progress**)
2. Create a foreman patch to accommodate these latest changes. NOTE: The updated plugin code will not work with the existing Foreman patch. This new patch will simplify implementations going forward.
3. Push a new release of the plugin(with new Netbox provider). Timing to be aligned with release date of the Foreman version that includes the above patch.
4. Handle authenticated? for different providers
5. Use validations from Proxy core
6. Add IPv6 address validation
7. Add more providers(i.e SolarWinds, Infoblox etc.)

In terms of adding more providers, is there one particular provider that is in demand right now? SolarWinds might be a good one to implement next, but let me know if there is any specific demand from the community, and I can go with that

Please have a look at the latest `develop` branch, and let me know what you think.

I look forward to your feedback!

**Comment by lzap** [comment]
> created a simple rake task

Nice idea.

> terms of adding more providers

I think having two providers is a great thing to have to prove the design is correct. I don't think you necessary need to add more at this point, remember more providers mean more support and maintainance. While code might look easy enough, setting up environments and other dance can be time consuming.

> Please have a look at the latest `develop` branch, and let me know what you think.

Lemme see and get back to you in this PR.

**Comment by lzap** [comment]
I think the code is good enough to be merged into smart-proxy now. I think good timing would be after we branch off 2.3 release in a month or so. Unless @ekohl has some concerns? Few nitpick tho:

I noticed that `next_ip` call returns a string directly (`next_ip.to_json`). It is usually good idea to always have a root element in XML/JSON/YAML or any other structured response because this allows you to make changes in the future. In this case I would suggest to have the response something like `{ "ip": "1.2.3.4" }`. At any point, you can extend the response without breaking your clients to something like (I am making this up) `{ "ip": "1.2.3.4", "warnings": ["Less than 10% of free IPs"] }`.

I just noticed that most of other endpoints actually returns a root element called "data". Except just one. Other than that the API feels just right, good job.

One more nitpic is that `ipam_helper.rb#errors` should have been (frozen) constant instead of method. This forces Ruby to allocate new hash over and over again.

It looks great! Stunning work and two providers - this is a dream. This can definitely go into smart proxy core already.

**Comment by grizzthedj** [comment]
@lzap I appreciate the feedback here. Good point about the root element for `next_ip`. The other nitpicks make sense as well. I will make these changes, as well as wait for any feedback from @ekohl.

**Comment by ekohl** [comment]
I do think the current version is a big improvement. There are some things that are surprising, though a lot of them could be qualified as nitpicks

* GET doesn't have a body so there is no `Content-Type`. Same goes for DELETE.
https://github.com/grizzthedj/smart_proxy_ipam/blob/f7494d94614a82235ba44cb1f29fc62f714c6931/lib/smart_proxy_ipam/api_resource.rb#L24
* Delete has an odd naming. The parameter is body, but it's really treated as a param.
https://github.com/grizzthedj/smart_proxy_ipam/blob/f7494d94614a82235ba44cb1f29fc62f714c6931/lib/smart_proxy_ipam/api_resource.rb#L31-L33
* Using the HTTP params to properly encode it would be good (just one example, repeated pattern):
https://github.com/grizzthedj/smart_proxy_ipam/blob/f7494d94614a82235ba44cb1f29fc62f714c6931/lib/smart_proxy_ipam/netbox/netbox_client.rb#L151
* halt is really for errors, you can also use `status`:
https://github.com/grizzthedj/smart_proxy_ipam/blob/f7494d94614a82235ba44cb1f29fc62f714c6931/lib/smart_proxy_ipam/ipam_api.rb#L330
* 200 is OK but you're not sending any content. HTTP 204 No Content is semantically more correct.
https://github.com/grizzthedj/smart_proxy_ipam/blob/f7494d94614a82235ba44cb1f29fc62f714c6931/lib/smart_proxy_ipam/ipam_api.rb#L378
* Is it really an array?
https://github.com/grizzthedj/smart_proxy_ipam/blob/f7494d94614a82235ba44cb1f29fc62f714c6931/lib/smart_proxy_ipam/ipam_api.rb#L36

What does stand out to me is that the IP cache should really be a singleton service. I think that right now there may an IP cache instance per HTTP request:
https://github.com/grizzthedj/smart_proxy_ipam/blob/f7494d94614a82235ba44cb1f29fc62f714c6931/lib/smart_proxy_ipam/netbox/netbox_client.rb#L25

For the singleton service, here is an example:
https://github.com/theforeman/smart-proxy/blob/9e94486a232f4eb1cb7e070ab833e9df9b07d118/modules/dhcp_isc/configuration_loader.rb#L28
https://github.com/theforeman/smart-proxy/blob/9e94486a232f4eb1cb7e070ab833e9df9b07d118/modules/dhcp_isc/dhcp_isc_plugin.rb#L18
https://github.com/theforeman/smart-proxy/blob/9e94486a232f4eb1cb7e070ab833e9df9b07d118/modules/dhcp_isc/configuration_loader.rb#L45

For future reference: you can open a PR from a branch in your own repo. Feel free to ping me if you want my opinion.

**Pull request closed without merging**
